### PR TITLE
KAFKA-8200: added Iterator methods for output to TopologyTestDriver

### PR DESCRIPTION
**Pull request by pkleindl**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java [hunk 1]
```diff
@@ -45,7 +45,6 @@
 import org.apache.kafka.streams.internals.WindowStoreFacade;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.PunctuationType;
-import org.apache.kafka.streams.processor.Punctuator;
```
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Why this change? Importing simplifies JavaDocs.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 2]
```diff
@@ -875,10 +876,11 @@ public void process(final String key, final Long value) {
         }
 
         private void flushStore() {
-            final KeyValueIterator<String, Long> it = store.all();
-            while (it.hasNext()) {
-                final KeyValue<String, Long> next = it.next();
-                context.forward(next.key, next.value);
+            try (final KeyValueIterator<String, Long> it = store.all()) {
```
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nice catch! Can we extract this into it's own PR so we can back port it to older branches?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. As we need a KIP, we should have a discussion there.

I am wondering, if it might be better to return an `Iterator` instead of `Iterable`? But we discuss on the KIP.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java (hunk 1 above)
Comment (pkleindl) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: It does, but it does not work well with "Organize imports". Which I usually do to avoided errors for unused imports.
The problem here is that Punctuator is never used in a javadoc link, therefore it is not recognized as a required import, at least not in my IDE.
So using the fully qualified name there seemed safer.
Any suggestion is welcome :-)

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 2 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Done https://github.com/apache/kafka/pull/6562

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java (hunk 1 above)
Comment (mjsax) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Seems to be shortcoming of your IDE. I don't have issue with IntelliJ. Maybe it's best to not do "Organize imports" and rely on the gradle checkstyle that will detect unused imports anyway.?

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 2 above)
Comment (mjsax) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Thanks.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java (hunk 1 above)
Comment (pkleindl) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Ok, I have reverted this part.

**Comment by cadonna** [comment]
After reading the discussion about `Iterable` vs. `Iterator` on the mailing list and looking at the code, I am wondering whether 

```
private final Map<String, Queue<ProducerRecord<byte[], byte[]>>> outputRecordsByTopic = new HashMap<>();
```

should  be changed to 

```
private final Map<String, Deque<ProducerRecord<byte[], byte[]>>> outputRecordsByTopic = new HashMap<>();
```
`Queue` does not guarantee iteration in FIFO-order whereas `Deque` does.

I guess, you could do this in this PR, since it seems a really small change to me, so it fits into the boyscout rule, i.e., "Leave your code better than you found it".

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 3]
```diff
@@ -942,21 +944,21 @@ public void close() {}
...
 
-        {
-            final TopologyTestDriver testDriver = new TopologyTestDriver(topology, config);
+
+        try (final TopologyTestDriver testDriver = new TopologyTestDriver(topology, config)) {
             Assert.assertNull(
-                "Closing the prior test driver should have cleaned up this store and value.",
-                testDriver.getKeyValueStore("storeProcessorStore").get("a")
+                    "Closing the prior test driver should have cleaned up this store and value.",
+                    testDriver.getKeyValueStore("storeProcessorStore").get("a")
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: The indentation should be four spaces instead of eight.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 4]
```diff
@@ -1088,4 +1090,66 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
         final TaskId taskId = new TaskId(0, 0);
         assertTrue(new File(appDir, taskId.toString()).exists());
     }
+
+    @Test
+    public void shouldReturnIteratorForMultipleRecords() {
+        testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
+
+        testDriver.pipeInput(consumerRecord1);
+        testDriver.pipeInput(consumerRecord1);
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Here it is better to use to distinct records, because if the code contains a bug that adds a record twice, you would not discover it.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 5]
```diff
@@ -1088,4 +1090,66 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        assertEquals(key1, outputRecord1.key());
+        assertEquals(value1, outputRecord1.value());
+        assertEquals(SINK_TOPIC_1, outputRecord1.topic());
+
+        final ProducerRecord outputRecord2 = output.next();
+
+        assertEquals(key1, outputRecord2.key());
+        assertEquals(value1, outputRecord2.value());
+        assertEquals(SINK_TOPIC_1, outputRecord2.topic());
+
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nit: remove empty line

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 6]
```diff
@@ -1088,4 +1090,66 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        assertEquals("Key1", outputRecord1.key());
+        assertEquals(12345L, outputRecord1.value());
+        assertEquals(SINK_TOPIC_1, outputRecord1.topic());
+
+        final ProducerRecord outputRecord2 = output.next();
+
+        assertEquals("Key2", outputRecord2.key());
+        assertEquals(6789L, outputRecord2.value());
+        assertEquals(SINK_TOPIC_1, outputRecord2.topic());
+
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nit: remove empty line

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 5 above)
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Add a check to verify, whether the iterator has no more elements.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 7]
```diff
@@ -1088,4 +1090,66 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        assertEquals("Key1", outputRecord1.key());
+        assertEquals(12345L, outputRecord1.value());
+        assertEquals(SINK_TOPIC_1, outputRecord1.topic());
+
+        final ProducerRecord outputRecord2 = output.next();
+
+        assertEquals("Key2", outputRecord2.key());
+        assertEquals(6789L, outputRecord2.value());
+        assertEquals(SINK_TOPIC_1, outputRecord2.topic());
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1157]: Add a check to verify, whether the iterator has no more elements.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 8]
```diff
@@ -1088,4 +1090,66 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        assertFalse(testDriver.iterateOutput(SINK_TOPIC_1).iterator().hasNext());
+
+    }
+
+    @Test
+    public void shouldReturnEmptyListForIterateOutputWithSerDes() {
+        testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
+
+        assertFalse(testDriver.iterateOutput(SINK_TOPIC_1, stringDeserializer, longDeserializer).iterator().hasNext());
+
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nit: remove empty line

**Review by cadonna — COMMENTED** [review]
Thank you @pkleindl for the PR!

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 8 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: @cadonna Hope I have addressed all your findings in the latest commit.

**Comment by pkleindl** [comment]
Retest this please.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 9]
```diff
@@ -1088,4 +1090,68 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        assertEquals(SINK_TOPIC_1, outputRecord2.topic());
+        assertFalse(output.hasNext());
+    }
+
+    @Test
+    public void shouldReturnIteratorForMultipleRecordsWithSerDes() {
+        testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
+
+        testDriver.pipeInput(recordFactory.create(SOURCE_TOPIC_1, "Key1", 12345L));
+        testDriver.pipeInput(recordFactory.create(SOURCE_TOPIC_1, "Key2", 6789L));
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Could you please extract the values `"Key1"`, `12345L`, `"Key2"`, and `"6789"` into variables and use the variables in the checks below? I am sorry, I missed that in my previous review.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 10]
```diff
@@ -1088,4 +1090,68 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
+        testDriver.pipeInput(consumerRecord2);
+
+        final Iterator<ProducerRecord<byte[], byte[]>> output = testDriver.iterateOutput(SINK_TOPIC_1).iterator();
+
+        final ProducerRecord outputRecord1 = output.next();
+
+        assertEquals(key1, outputRecord1.key());
+        assertEquals(value1, outputRecord1.value());
+        assertEquals(SINK_TOPIC_1, outputRecord1.topic());
+
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1115]: FYI: There is the static nested class `Record` in `TopologyTestDriverTest`, that can be used to compare records.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 11]
```diff
@@ -1088,4 +1090,68 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
         final TaskId taskId = new TaskId(0, 0);
         assertTrue(new File(appDir, taskId.toString()).exists());
     }
+
+    @Test
+    public void shouldReturnIteratorForMultipleRecords() {
+        testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
+
+        testDriver.pipeInput(consumerRecord1);
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nit: I would also create a local `consumerRecord1` as you did with `consumerRecord2`. Alternatively, you could use the global `consumerRecord2` instead of creating a local one. In my opinion, mixing the global and local makes the code less readable.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 10 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1115]: I don't see how I could use that here, as far as I can see it is used to compare the records from the MockProcessor to the input records.
In my case I would have to compare input to output records which don't match on the topic name.
Or am I missing something?

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 9 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Done

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 11 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Done
I took the liberty and renamed the methods to "iterable" (from "iterate") because I missed that on the name-change.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 10 above)
Comment (cadonna) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:1115]: Sorry, you are right! My bad!

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 12]
```diff
@@ -1092,17 +1092,19 @@ public void shouldCreateStateDirectoryForStatefulTopology() {
...
 
     @Test
-    public void shouldReturnIteratorForMultipleRecords() {
+    public void shouldReturnIterableForMultipleRecords() {
         testDriver = new TopologyTestDriver(setupSourceSinkTopology(), config);
-
-        testDriver.pipeInput(consumerRecord1);
+        final byte[] key1 = "key1".getBytes();
+        final byte[] value1 = "value1".getBytes();
+        final ConsumerRecord<byte[], byte[]> consumerRecord1 = consumerRecordFactory.create(SOURCE_TOPIC_1, key1, value1, new RecordHeaders(), 0L);
```
Comment (cadonna) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: nit: we usually try to keep lines under 120 characters.

**Review by cadonna — APPROVED** [review]
LGTM

Only one last nit.

Thank you @pkleindl !

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 12 above)
Comment (pkleindl) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Done, my IDE settings were off regarding wrapping

**Comment by cadonna** [comment]
Failure of org.apache.kafka.streams.integration.ResetIntegrationTest.shouldNotAllowToResetWhenInputTopicAbsent not related

Retest this, please

**Comment by cadonna** [comment]
Failure of kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
unrelated

Retest this, please

**Comment by pkleindl** [comment]
@cadonna @mjsax I will close this as discussed on the mailing list in favor of KIP-456
Thanks a lot for your help, it was still good for some PR baby steps :-)
